plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/DecimalCoercers.java

```diff
@@ -286,6 +292,117 @@ protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int pos
+                minValue = Short.MIN_VALUE;
+                maxValue = Short.MAX_VALUE;
+            }
+            else if (toType.equals(INTEGER)) {
+                minValue = Integer.MIN_VALUE;
+                maxValue = Integer.MAX_VALUE;
+            }
+            else if (toType.equals(BIGINT)) {
+                minValue = Long.MIN_VALUE;
+                maxValue = Long.MAX_VALUE;
+            }
+            else {
+                throw new TrinoException(NOT_SUPPORTED, format("Could not create Coercer from Decimal to %s", toType));
+            }
+        }
+
+        @Override
+        protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int position)
+        {
+            try {
+                String stringValue = Decimals.toString((Int128) fromType.getObject(block, position), fromType.getScale());
+                int dotPosition = stringValue.indexOf(".");
+                long longValue = Long.parseLong(stringValue.substring(0, dotPosition > 0 ? dotPosition : stringValue.length()));
+                // Hive truncates digits (also before the decimal point), which can be perceived as a bug
+                if (longValue > maxValue || longValue < minValue) {
+                    blockBuilder.appendNull();
+                }
+                else {
+                    toType.writeLong(blockBuilder, longValue);
+                }
+            }
+            catch (NumberFormatException e) {
+                blockBuilder.appendNull();
+            }
+        }
+    }
+
+    private static class ShortDecimalToIntegerCoercer<T extends Type>
+            extends TypeCoercer<DecimalType, T>
+    {
+        private final long minValue;
+        private final long maxValue;
+
+        public ShortDecimalToIntegerCoercer(DecimalType fromType, T toType)
+        {
+            super(fromType, toType);
+
+            if (toType.equals(TINYINT)) {
+                minValue = Byte.MIN_VALUE;
+                maxValue = Byte.MAX_VALUE;
+            }
+            else if (toType.equals(SMALLINT)) {
+                minValue = Short.MIN_VALUE;
+                maxValue = Short.MAX_VALUE;
+            }
+            else if (toType.equals(INTEGER)) {
+                minValue = Integer.MIN_VALUE;
+                maxValue = Integer.MAX_VALUE;
+            }
+            else if (toType.equals(BIGINT)) {
+                minValue = Long.MIN_VALUE;
+                maxValue = Long.MAX_VALUE;
+            }
+            else {
+                throw new TrinoException(NOT_SUPPORTED, format("Could not create Coercer from Decimal to %s", toType));
+            }
+        }
+
+        @Override
+        protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int position)
+        {
+            try {
+                String stringValue = Decimals.toString(fromType.getLong(block, position), fromType.getScale());
+                long longValue = Long.parseLong(stringValue.substring(0, stringValue.indexOf(".")));
```

Comment:
Same as above - limit the handling of `NumberFormatException` only for this line.